src/file_pipe_log/pipe.rs

```diff
@@ -505,5 +669,288 @@ mod tests {
+            purge_threshold: ReadableSize::mb(1),
+            enable_log_recycle: true,
+            format_version: Version::V2,
+            ..Default::default()
+        };
+        let queue = LogQueue::Append;
+
+        let pipe_log = new_test_pipes(&cfg).unwrap();
+        assert_eq!(pipe_log.file_span(queue), (1, 1));
+
+        let header_size = LogFileFormat::len() as u64;
+
+        // generate file 1, 2, 3
+        let content: Vec<u8> = vec![b'a'; 1024];
+        let file_handle = pipe_log.append(queue, &content).unwrap();
+        pipe_log.maybe_sync(queue, false).unwrap();
+        assert_eq!(file_handle.id.seq, 1);
+        assert_eq!(file_handle.offset, header_size);
+        assert_eq!(pipe_log.file_span(queue).1, 2);
+
+        let file_handle = pipe_log.append(queue, &content).unwrap();
+        pipe_log.maybe_sync(queue, false).unwrap();
+        assert_eq!(file_handle.id.seq, 2);
+        assert_eq!(file_handle.offset, header_size);
+        assert_eq!(pipe_log.file_span(queue).1, 3);
+
+        // purge file 1
+        assert_eq!(pipe_log.purge_to(FileId { queue, seq: 2 }).unwrap(), 1);
+        assert_eq!(pipe_log.file_span(queue).0, 2);
+
+        // cannot purge active file
+        assert!(pipe_log.purge_to(FileId { queue, seq: 4 }).is_err());
+
+        // append position
+        let s_content = b"short content".to_vec();
+        let file_handle = pipe_log.append(queue, &s_content).unwrap();
+        pipe_log.maybe_sync(queue, false).unwrap();
+        assert_eq!(file_handle.id.seq, 3);
+        assert_eq!(file_handle.offset, header_size);
+
+        let file_handle = pipe_log.append(queue, &s_content).unwrap();
+        pipe_log.maybe_sync(queue, false).unwrap();
+        assert_eq!(file_handle.id.seq, 3);
+        assert_eq!(
+            file_handle.offset,
+            header_size as u64 + s_content.len() as u64
+        );
+
+        let content_readed = pipe_log
+            .read_bytes(FileBlockHandle {
+                id: FileId { queue, seq: 3 },
+                offset: header_size as u64,
+                len: s_content.len(),
+            })
+            .unwrap();
+        assert_eq!(content_readed, s_content);
+        // try to fetch abnormal entry
+        let abnormal_content_readed = pipe_log.read_bytes(FileBlockHandle {
+            id: FileId { queue, seq: 12 }, // abnormal seq
+            offset: header_size as u64,
+            len: s_content.len(),
+        });
+        assert!(abnormal_content_readed.is_err());
+
+        // leave only 1 file to truncate
+        assert!(pipe_log.purge_to(FileId { queue, seq: 3 }).is_ok());
+        assert_eq!(pipe_log.file_span(queue), (3, 3));
+
+        // fetch active file
+        let file_context = pipe_log.fetch_active_file(LogQueue::Append).unwrap();
+        assert_eq!(file_context.version, Version::V2);
+        assert_eq!(file_context.id.seq, 3);
+    }
+
+    #[test]
+    #[cfg(feature = "failpoints")]
```

Comment:
Move all failpoint tests to its dedicated crate. Failpoint tests must be run with test-thread=1.